### PR TITLE
Handle CTU diagnostic macro expansions better.

### PR DESCRIPTION
**Pull request by balazske**
If a macro is to be expanded during diagnostic report creation the preprocessor for the original source file (where the macro was imported from) is used instead of the preprocessor for the actual source file (that do not contain correct macro informations). (If the macro is in imported code.)

### CODE REVIEW

**Comment by Szelethus** [comment]
Wow, I would never have guessed it's possible to retrieve the `Preprocessor` for another AST!

I glanced over the patch and I like what I'm seeing, though it would be great to test some actual macro expansions, something like this:

https://github.com/Ericsson/clang/blob/ctu-clang7/test/Analysis/plist-macros-with-expansion.cpp

**Comment by balazske** [comment]
Added a test that fails without the fix with the "unsortable locations found" error. The test fails too if the original solution in `isBeforeInTranslationUnit` is used.

**Comment by Szelethus** [comment]
Do you refer to this patch?

https://reviews.llvm.org/D59934

**Comment by balazske** [comment]
Yes, that fix is not correct because it do not find macros in other TU (that has different SourceManager and different macro history that is not imported at AST import).

**Comment by Szelethus** [comment]
Hmm, I see. Then which patch does this depend on?
What does failure of the lit tests mean (segfault, assertion failure, incorrect output)?

File: include/clang/CrossTU/CrossTranslationUnit.h [hunk 1]
```diff
@@ -160,6 +160,16 @@ class CrossTranslationUnitContext {
...
   void emitCrossTUDiagnostics(const IndexError &IE);
 
+  /// Determine the original source location in the original TU for an
+  /// imported source location.
+  /// \p ToLoc Source location in the imported-to AST.
+  /// \p FromLoc Returned source location in the imported-from AST.
+  /// \return The ASTUnit for the source file of FromLoc. If the ToLoc is not an
+  /// imported location or other error occurs, nullptr is returned.
+  clang::ASTUnit *
+  GetImportedFromSourceLocation(const clang::SourceLocation &ToLoc,
```
Comment (martong) [inline include/clang/CrossTU/CrossTranslationUnit.h]: I think, using Expected could be better for error handling and for code comprehension:
`Expected<tuple<ASTUnit, SourceLocation /*FromLoc*/>> GetImportedFromSourceLocation(const clang::SourceLocation &ToLoc)`

File: include/clang/StaticAnalyzer/Core/PathDiagnosticConsumers.h [hunk 2]
```diff
@@ -21,17 +21,19 @@ namespace clang {
...
 
 class PathDiagnosticConsumer;
 typedef std::vector<PathDiagnosticConsumer*> PathDiagnosticConsumers;
 
-#define ANALYSIS_DIAGNOSTICS(NAME, CMDFLAG, DESC, CREATEFN)\
-void CREATEFN(AnalyzerOptions &AnalyzerOpts,\
-              PathDiagnosticConsumers &C,\
-              const std::string &Prefix,\
-              const Preprocessor &PP);
+#define ANALYSIS_DIAGNOSTICS(NAME, CMDFLAG, DESC, CREATEFN)                    \
```
Comment (martong) [inline include/clang/StaticAnalyzer/Core/PathDiagnosticConsumers.h:33]: Please avoid unrelated changes.

File: include/clang/CrossTU/CrossTranslationUnit.h [hunk 3]
```diff
@@ -168,6 +178,7 @@ class CrossTranslationUnitContext {
 
   llvm::StringMap<std::unique_ptr<clang::ASTUnit>> FileASTUnitMap;
   llvm::StringMap<clang::ASTUnit *> FunctionASTUnitMap;
+  llvm::StringMap<clang::ASTUnit *> SrcFileASTUnitMap;
```
Comment (martong) [inline include/clang/CrossTU/CrossTranslationUnit.h]: Why can't we use `FileASTUnitMap`?
Please include the answer as a comment above both member variables: `FileASTUnitMap` and `SrcFileASTUnitMap`.

**Review by martong — COMMENTED** [review]
Nice work!

**Comment by martong** [comment]
@Xazax-hun Adding you as a reviewer, because this involves changes in CTU files too and I'd like you to get a glimpse on this before we upstream.

**Comment by martong** [comment]
run tests (this comment is intended for the jenkins bot)

File: lib/CrossTU/CrossTranslationUnit.cpp [hunk 4]
```diff
@@ -349,6 +349,7 @@ llvm::Expected<ASTUnit *> CrossTranslationUnitContext::loadExternalAST(
           ASTUnit::LoadEverything, Diags, CI.getFileSystemOpts()));
       Unit = LoadedUnit.get();
       FileASTUnitMap[ASTFileName] = std::move(LoadedUnit);
+      SrcFileASTUnitMap[Unit->getOriginalSourceFileName()] = Unit;
```
Comment (Xazax-hun) [inline lib/CrossTU/CrossTranslationUnit.cpp]: I wonder if this would work well if we have different files (on different paths) with the same name. Shouldn't we make sure this is an absolute path for safety?

File: include/clang/StaticAnalyzer/Core/PathDiagnosticConsumers.h (hunk 2 above)
Comment (balazske) [reply include/clang/StaticAnalyzer/Core/PathDiagnosticConsumers.h:33]: The whole macro ANALYSIS_DIAGNOSTICS was changed, and clang-format updated the formatting for it. (Other option is to add new things with the existing "bad" formatting.)

File: lib/CrossTU/CrossTranslationUnit.cpp (hunk 4 above)
Comment (balazske) [reply lib/CrossTU/CrossTranslationUnit.cpp]: If there are different files with identical name that does not work. The assumption is that this "original filename" is the same that is contained in the `SourceLocation` for the main (in that `ASTUnit`) AST file. An assert can be added here to check that the map has no entry already.

File: include/clang/CrossTU/CrossTranslationUnit.h (hunk 3 above)
Comment (balazske) [reply include/clang/CrossTU/CrossTranslationUnit.h]: It can be possible to get the source filename from AST filename but it is more complicated and may not work in all cases or if the AST file naming is changed later.

**Comment by balazske** [comment]
If the fix is omitted the test fails with:
```
Unsortable locations found
UNREACHABLE executed at llvm/tools/clang/lib/Basic/SourceManager.cpp:2040!
```
If the other fix (in `isBeforeInTranslationUnit`) is used we get incorrect output: Some macros are not expanded (that are defined in the to-be-imported file only) (I think it is possible to get wrongly expanded macros if there is a macro with same name in both files).

File: include/clang/CrossTU/CrossTranslationUnit.h (hunk 3 above)
Comment (martong) [reply include/clang/CrossTU/CrossTranslationUnit.h]: OK

File: include/clang/StaticAnalyzer/Core/PathDiagnosticConsumers.h (hunk 2 above)
Comment (martong) [reply include/clang/StaticAnalyzer/Core/PathDiagnosticConsumers.h:33]: As I see, we just added a new parameter, but previously I missed that. So, OK.

**Review by martong — COMMENTED** [review]
Looks good to me! However, the PR job had failed: http://ci1.codechecker.gic.ericsson.se:8080/job/ctu_pipeline_pullrequest_clang8/18/ could you take a look on that?

**Comment by balazske** [comment]
The test fails because there is still a problem: Only source locations are handled that are in the "main" source file in a TU but not those that are in headers.

File: include/clang/AST/ASTImporterSharedState.h [hunk 5]
```diff
@@ -16,18 +16,24 @@
...
 
+class ASTUnit;
 class TranslationUnitDecl;
 
 /// Importer specific state, which may be shared amongst several ASTImporter
 /// objects.
 class ASTImporterSharedState {
+public:
+  using ImportedFileIDMap =
+      llvm::DenseMap<FileID, std::pair<FileID, ASTUnit *>>;
```
Comment (martong) [inline include/clang/AST/ASTImporterSharedState.h:34]: Why must we store this info in the shared state? Could you please elaborate?

Comment (balazske) [reply include/clang/AST/ASTImporterSharedState.h:34]: The goal is to get the `ASTUnit` from a `FileID` that was imported, by only knowing the `FileID` value (not the `ASTImporter` or `ASTUnit`). This map can be used to get the ASTUnit if there are multiple ASTImporters (and corresponding ASTUnits). (Previously the filename of the source file was used to get the ASTUnit in `CrossTranslationUnitContext` but this does not work if the file is an include file. And using the filename can be source of errors.)

Comment (martong) [reply include/clang/AST/ASTImporterSharedState.h:34]: Why this can't be a member of the ASTImporter class? Why would it be a problem if the different ASTImporter objects have their own copy?

Comment (balazske) [reply include/clang/AST/ASTImporterSharedState.h:34]: This map is needed to find out the `ASTUnit` in `CrossTranslationUnitContext::GetImportedFromSourceLocation`. It should contain `FileID`'s imported by any of the `ASTImporter`'s and the data is used independent of a single `ASTImporter`. Every `ASTImporter` inserts its own `ASTUnit` into it (this is why the `ASTUnit` is added to `ASTImporter`: The `ASTImporter` can get the `FileID` mapping and at the same time the `ASTUnit`). (If every `ASTImporter` had its own copy we can loop over every `ASTUnit` to find the matching one. This can be an alternative solution, then a simple FileID->FileID map is sufficient in `ASTImporter`, and no store of `ASTUnit`. But the lookup becomes slower.)

File: include/clang/AST/ASTImporter.h [hunk 6]
```diff
@@ -132,6 +133,9 @@ class TypeSourceInfo;
     /// The file managers we're importing to and from.
     FileManager &ToFileManager, &FromFileManager;
 
+    /// The ASTUnit that this importer belongs to, if any.
+    ASTUnit *Unit;
```
Comment (martong) [inline include/clang/AST/ASTImporter.h:137]: Unit->getASTContext() should return a ref which is equivalent to `ToContext`, right?
Then perhaps we should assert on this in the ctor.
Plus, I think the comment should indicate that this is associated with the "to" context.

File: include/clang/AST/ASTImporterSharedState.h (hunk 5 above)
Comment (martong) [reply include/clang/AST/ASTImporterSharedState.h:34]: Okay, could you please then add a comment to the member which describes this (also mentioning the alternative and why that would be slower).

File: include/clang/AST/ASTImporter.h (hunk 6 above)
Comment (balazske) [reply include/clang/AST/ASTImporter.h:137]: This is the `FromContext`, not the To. The ASTImporter's in `CrossTranslationUnitContext` have all the same `ToContext` (and same To AST) and a different `FromContext` (that belongs to the loaded From AST). The `ASTUnit` objects manage the loaded From AST's.

Comment (martong) [reply include/clang/AST/ASTImporter.h:137]: Okay, then we should assert that Unit->getASTContext() should return a ref which is equivalent to FromContext

Comment (balazske) [reply include/clang/AST/ASTImporter.h:137]: I do not like to put assert into `ASTImporter` for this, the passed `ASTUnit` pointer is just an "arbitrary value" here that is put into ImportedFileIDs (working of the ASTImporter itself does not depend on any value of it). (And only for this assert the ASTUnit header must be used.) The assert can be put into where the ASTImporter is created.

**Review by martong — APPROVED** [review]

**Comment by balazske** [comment]
First part:
https://reviews.llvm.org/D64554

**Comment by balazske** [comment]
https://reviews.llvm.org/D64635
https://reviews.llvm.org/D64638
